src/iface/fragmentation.rs

```diff
@@ -0,0 +1,555 @@
+    },
+}
+
+impl<'a, Info: PacketAssemblerInfo> PacketAssembler<'a, Info> {
+    /// Create a new empty buffer for fragments.
+    pub fn new<S>(storage: S) -> Self
+    where
+        S: Into<ManagedSlice<'a, u8>>,
+    {
+        let s = storage.into();
+        PacketAssembler {
+            buffer: s,
+            assembler: AssemblerState::NotInit,
+        }
+    }
+
+    /// Start with saving fragments.
+    /// We initialize the assembler with the total size of the final packet.
+    ///
+    /// # Errors
+    ///
+    /// - Returns [`Error::PacketAssemblerBufferTooSmall`] when the buffer is too smal for holding all the
+    /// fragments of a packet.
+    pub(crate) fn start(
+        &mut self,
+        total_size: usize,
+        info: Info,
+        start_time: Instant,
+    ) -> Result<()> {
+        match &mut self.buffer {
+            ManagedSlice::Borrowed(b) if b.len() < total_size => {
+                return Err(Error::PacketAssemblerBufferTooSmall);
+            }
+            ManagedSlice::Borrowed(_) => (),
+            #[cfg(any(feature = "std", feature = "alloc"))]
+            ManagedSlice::Owned(b) => {
+                b.resize(total_size, 0);
+            }
+        }
+
+        self.assembler = AssemblerState::Assembling {
+            assembler: Assembler::new(total_size),
+            total_size,
+            info,
+            last_updated: start_time,
+            started_on: start_time,
+        };
+
+        Ok(())
+    }
+
+    /// Add a fragment into the packet that is being reassembled.
+    ///
+    /// # Errors
+    ///
+    /// - Returns [`Error::PacketAssemblerNotInit`] when the assembler was not initialized (try initializing the
+    /// assembler with [Self::start]).
+    /// - Returns [`Error::PacketAssemblerBufferTooSmall`] when trying to add data into the buffer at a non-existing
+    /// place.
+    /// - Returns [`Error::PacketAssemblerOverlap`] when there was an overlap when adding data.
+    pub(crate) fn add(&mut self, data: &[u8], offset: usize, now: Instant) -> Result<bool> {
+        match self.assembler {
+            AssemblerState::NotInit => Err(Error::PacketAssemblerNotInit),
+            AssemblerState::Assembling {
+                ref mut assembler,
+                total_size,
+                ref info,
+                ref mut last_updated,
+                ..
+            } => {
+                let offset = info.calc_offset(offset);
+
+                if offset + data.len() > total_size {
+                    return Err(Error::PacketAssemblerBufferTooSmall);
+                }
+
+                let len = data.len();
+                self.buffer[offset..][..len].copy_from_slice(data);
+
+                match assembler.add(offset, data.len()) {
+                    Ok(false) => {
+                        *last_updated = now;
+                        self.is_complete()
+                    }
+                    Ok(true) => Err(Error::PacketAssemblerOverlap),
```

Comment:
is it required to drop packets with overlap? AFAIK it's not for IPv4 fragmentation. With that, we could avoid making changes to `Assembler`

Reply:
Changed it.